servers/zms/src/main/java/com/yahoo/athenz/zms/DBService.java

```diff
@@ -2349,6 +2392,42 @@ Policy updateTemplatePolicy(Policy policy, String domainName, String policyName,
         return templatePolicy;
     }
     
+    ServiceIdentity updateTemplateServiceIdentity(ServiceIdentity serviceIdentity, String domainName, String serviceIdentityName, List<TemplateParam> params) {
+        
+        String templateServiceName = serviceIdentityName;
+        if (params != null) {
+            for (TemplateParam param : params) {
+                final String paramKey = "_" + param.getName() + "_";
+                System.out.println("param key &&&&&&&&&& " + paramKey + " &&&&&&");
+                templateServiceName = templateServiceName.replace(paramKey, param.getValue());
+            }
+        }
+        
+        ServiceIdentity templateServiceIdentity = new ServiceIdentity().setName(ZMSUtils.serviceResourceName(domainName, serviceIdentityName));
+        
+        templateServiceIdentity.setDescription(serviceIdentity.getDescription());
+        templateServiceIdentity.setExecutable(serviceIdentity.getExecutable());
+        templateServiceIdentity.setGroup(serviceIdentity.getGroup());
+        templateServiceIdentity.setUser(serviceIdentity.getUser());
+        templateServiceIdentity.setProviderEndpoint(serviceIdentity.getProviderEndpoint());
+        
+        List<PublicKeyEntry> publicKeyEntries = serviceIdentity.getPublicKeys();
+        List<PublicKeyEntry> newPublicKeyEntries = new ArrayList<PublicKeyEntry>();
+        if (publicKeyEntries != null && !publicKeyEntries.isEmpty()) {
+            for (PublicKeyEntry publicKeyEntry : publicKeyEntries) {
+                PublicKeyEntry newPublicKeyEntry = new PublicKeyEntry();
+                newPublicKeyEntry.setId(publicKeyEntry.getId());
+                newPublicKeyEntry.setKey(publicKeyEntry.getKey());
+                newPublicKeyEntries.add(newPublicKeyEntry);
+            }
+        }
+        templateServiceIdentity.setPublicKeys(newPublicKeyEntries);
+        
+        templateServiceIdentity.setHosts(serviceIdentity.getHosts());
```

Comment:
we shouldn't set a reference to the host list since it could be modified. Instead we should do:
if (hosts != null) {
templateServiceIdentity.setHosts(new ArrayList(hosts));
}